File: JavaScript/6js 2/nbMonths.js
/*
|--------------------------------------------------------------------------
| Buying a car
|--------------------------------------------------------------------------
|
*/

function nbMonths(spo, spn, spm, plm) {
   let months = 0;
   let savings = 0;
   while (spo + savings < spn) {
      months++;
      if (months % 2 === 0) plm += 0.5;
      spo *= 1 - plm / 100;
      spn *= 1 - plm / 100;
      savings += spm;
   }
   return [months, Math.round(spo + savings - spn)];
}


const f = (nbMonths = (
   o,
   n,
   s,
   l,
   b = 0,
   e = 0,
   d = 0,
   r = b + o - n,
   k = e ? l + 0.5 : l,
   m = 1 - k / 100
) => (r < 0 ? f(o * m, n * m, s, k, b + s, !e, d + 1) : [d, (r + 0.5) ^ 0]));


function nbMonths(myCar, newCar, perMonth, loss) {
   loss = 1 - loss / 100;
   let w = myCar;
   for (let m = 0; w < newCar; m++) {
      loss -= m % 2 == 0 ? 0 : 0.005;
      myCar *= loss;
      newCar *= loss;
      w = myCar + perMonth * (m + 1);
   }
   return [m, Math.round(w - newCar)];
}


console.log(nbMonths(2000, 8000, 1000, 1.5), [6, 766]);
console.log(nbMonths(12000, 8000, 1000, 1.5), [0, 4000]);

File: Python/6py 2/nb_months.py
"""
|--------------------------------------------------------------------------
| Buying a car
|--------------------------------------------------------------------------
|
"""

def nb_months(spo, spn, spm, plm):
   months, savings = 0, 0
   while spo + savings < spn:
      months += 1
      if months % 2 == 0: plm += 0.5
      spo *= 1 - plm / 100
      spn *= 1 - plm / 100
      savings += spm
   return [months, round(spo + savings - spn)]


def nbMonths(p0, p1, s, r, m = 0, t = 0):
   if p1 <= p0 + t: return [m, round(p0 - p1 + t)]
   p0 -= p0 * .01 * r
   p1 -= p1 * .01 * r
   t += s
   m += 1
   if m % 2 == 1: r += .5
   return nbMonths(p0, p1, s, r, m, t)


def nbMonths(old, new, savings, percent, month=0):
   if month*savings + old > new: return [month, round((month*savings + old) - new)]
   if month and month%2: percent += 0.5
   return nbMonths(old*(1-percent/100), new*(1-percent/100), savings, percent, month+1)


print(nb_months(2000, 8000, 1000, 1.5), [6, 766])
print(nb_months(12000, 8000, 1000, 1.5) ,[0, 4000])
print(nb_months(8000, 12000, 500, 1), [8, 597])
print(nb_months(18000, 32000, 1500, 1.25), [8, 332])
print(nb_months(7500, 32000, 300, 1.55), [25, 122])